// openmemory-js/dist/ai/mcp_tools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolRegistry = void 0;
const zod_1 = require("zod");
const zod_to_json_schema_1 = require("zod-to-json-schema");
const types_js_1 = require("@modelcontextprotocol/sdk/types.js");
class ToolRegistry {
    tools = new Map();
    tool(name, description, inputSchema, callback) {
        this.tools.set(name, {
            name,
            description,
            inputSchema: zod_1.z.object(inputSchema),
            callback
        });
    }
    apply(server) {
        const srv = server.server;
        srv.setRequestHandler(types_js_1.ListToolsRequestSchema, async () => {
            return {
                tools: Array.from(this.tools.values()).map(t => {
                    const jsonSchema = (0, zod_to_json_schema_1.zodToJsonSchema)(t.inputSchema, {
                        target: "jsonSchema2019-09"
                    });
                    if (jsonSchema && typeof jsonSchema === 'object') {
                        delete jsonSchema.$schema;
                    }
                    return {
                        name: t.name,
                        description: t.description,
                        inputSchema: jsonSchema
                    };
                })
            };
        });
        srv.setRequestHandler(types_js_1.CallToolRequestSchema, async (req, extra) => {
            const name = req.params.name;
            const tool = this.tools.get(name);
            if (!tool) {
                throw new types_js_1.McpError(types_js_1.ErrorCode.MethodNotFound, `Tool not found: ${name}`);
            }
            const args = req.params.arguments || {};
            const parse = await tool.inputSchema.safeParseAsync(args);
            if (!parse.success) {
                throw new types_js_1.McpError(types_js_1.ErrorCode.InvalidParams, `Invalid arguments: ${parse.error.message}`);
            }
            try {
                return await tool.callback(parse.data, extra);
            }
            catch (err) {
                return {
                    content: [{ type: "text", text: `Error: ${err.message}` }],
                    isError: true
                };
            }
        });
    }
}
exports.ToolRegistry = ToolRegistry;
