// openmemory-js/dist/core/vector/postgres.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostgresVectorStore = void 0;
const embed_1 = require("../../memory/embed");
class PostgresVectorStore {
    db;
    table;
    usePgVector;
    constructor(db, tableName = "vectors", usePgVector = false) {
        this.db = db;
        this.table = tableName;
        this.usePgVector = usePgVector;
        console.error(`[PostgresVectorStore] mode: ${usePgVector ? 'pgvector (native)' : 'sqlite (compat)'}`);
    }
    async storeVector(id, sector, vector, dim, user_id) {
        console.error(`[Vector] Storing ID: ${id}, Sector: ${sector}, Dim: ${dim}`);
        if (this.usePgVector) {
            const v_str = JSON.stringify(vector);
            const sql = `insert into ${this.table}(id,sector,user_id,v,dim) values($1,$2,$3,$4::vector,$5) on conflict(id,sector) do update set user_id=excluded.user_id,v=excluded.v,dim=excluded.dim`;
            await this.db.run_async(sql, [id, sector, user_id || "anonymous", v_str, dim]);
        }
        else {
            const v = (0, embed_1.vectorToBuffer)(vector);
            const sql = `insert into ${this.table}(id,sector,user_id,v,dim) values($1,$2,$3,$4,$5) on conflict(id,sector) do update set user_id=excluded.user_id,v=excluded.v,dim=excluded.dim`;
            await this.db.run_async(sql, [id, sector, user_id || "anonymous", v, dim]);
        }
    }
    async deleteVector(id, sector) {
        await this.db.run_async(`delete from ${this.table} where id=$1 and sector=$2`, [id, sector]);
    }
    async deleteVectors(id) {
        await this.db.run_async(`delete from ${this.table} where id=$1`, [id]);
    }
    async searchSimilar(sector, queryVec, topK, user_id) {
        if (this.usePgVector) {
            const v_str = JSON.stringify(queryVec);
            let filter_sql = "where sector = $2";
            const args = [v_str, sector, topK];
            if (user_id) {
                filter_sql += " and user_id = $4";
                args.push(user_id);
            }
            const sql = `
                select id, 1 - (v <=> $1::vector) as similarity
                from ${this.table}
                ${filter_sql}
                order by v <=> $1::vector
                limit $3
            `;
            const rows = await this.db.all_async(sql, args);
            console.error(`[Vector] pgvector search in sector: ${sector}${user_id ? `, user: ${user_id}` : ''}, returned ${rows.length} results`);
            return rows.map(r => ({ id: r.id, score: r.similarity }));
        }
        else {
            let filter_sql = "where sector=$1";
            const args = [sector];
            if (user_id) {
                filter_sql += " and user_id=$2";
                args.push(user_id);
            }
            const rows = await this.db.all_async(`select id,v,dim from ${this.table} ${filter_sql}`, args);
            console.error(`[Vector] sqlite-compat search in sector: ${sector}${user_id ? `, user: ${user_id}` : ''}, found ${rows.length} rows`);
            const sims = [];
            for (const row of rows) {
                const vec = (0, embed_1.bufferToVector)(row.v);
                const sim = (0, embed_1.cosineSimilarity)(queryVec, vec);
                sims.push({ id: row.id, score: sim });
            }
            sims.sort((a, b) => b.score - a.score);
            return sims.slice(0, topK);
        }
    }
    async getVector(id, sector) {
        if (this.usePgVector) {
            const row = await this.db.get_async(`select v::text as v_txt,dim from ${this.table} where id=$1 and sector=$2`, [id, sector]);
            if (!row)
                return null;
            return { vector: JSON.parse(row.v_txt), dim: row.dim };
        }
        else {
            const row = await this.db.get_async(`select v,dim from ${this.table} where id=$1 and sector=$2`, [id, sector]);
            if (!row)
                return null;
            return { vector: (0, embed_1.bufferToVector)(row.v), dim: row.dim };
        }
    }
    async getVectorsById(id) {
        if (this.usePgVector) {
            const rows = await this.db.all_async(`select sector,v::text as v_txt,dim from ${this.table} where id=$1`, [id]);
            return rows.map(row => ({ sector: row.sector, vector: JSON.parse(row.v_txt), dim: row.dim }));
        }
        else {
            const rows = await this.db.all_async(`select sector,v,dim from ${this.table} where id=$1`, [id]);
            return rows.map(row => ({ sector: row.sector, vector: (0, embed_1.bufferToVector)(row.v), dim: row.dim }));
        }
    }
    async getVectorsBySector(sector) {
        if (this.usePgVector) {
            const rows = await this.db.all_async(`select id,v::text as v_txt,dim from ${this.table} where sector=$1`, [sector]);
            return rows.map(row => ({ id: row.id, vector: JSON.parse(row.v_txt), dim: row.dim }));
        }
        else {
            const rows = await this.db.all_async(`select id,v,dim from ${this.table} where sector=$1`, [sector]);
            return rows.map(row => ({ id: row.id, vector: (0, embed_1.bufferToVector)(row.v), dim: row.dim }));
        }
    }
}
exports.PostgresVectorStore = PostgresVectorStore;
