// openmemory-js/dist/core/vector/valkey.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValkeyVectorStore = void 0;
const ioredis_1 = __importDefault(require("ioredis"));
const cfg_1 = require("../cfg");
const embed_1 = require("../../memory/embed");
class ValkeyVectorStore {
    client;
    constructor() {
        this.client = new ioredis_1.default({
            host: cfg_1.env.valkey_host || "localhost",
            port: cfg_1.env.valkey_port || 6379,
            password: cfg_1.env.valkey_password,
        });
    }
    getKey(id, sector) {
        return `vec:${sector}:${id}`;
    }
    async storeVector(id, sector, vector, dim, user_id) {
        const key = this.getKey(id, sector);
        const buf = (0, embed_1.vectorToBuffer)(vector);
        await this.client.hset(key, {
            v: buf,
            dim: dim,
            user_id: user_id || "anonymous",
            id: id,
            sector: sector
        });
    }
    async deleteVector(id, sector) {
        const key = this.getKey(id, sector);
        await this.client.del(key);
    }
    async deleteVectors(id) {
        let cursor = "0";
        do {
            const res = await this.client.scan(cursor, "MATCH", `vec:*:${id}`, "COUNT", 100);
            cursor = res[0];
            const keys = res[1];
            if (keys.length)
                await this.client.del(...keys);
        } while (cursor !== "0");
    }
    async searchSimilar(sector, queryVec, topK, user_id) {
        // Valkey/Redis doesn't support user_id filtering in FT.SEARCH easily
        // For now we'll need to post-filter or use a more complex query
        const indexName = `idx:${sector}`;
        const blob = (0, embed_1.vectorToBuffer)(queryVec);
        try {
            // Use FT.SEARCH with vector similarity
            const res = await this.client.call("FT.SEARCH", indexName, `*=>[KNN ${topK * 2} @v $blob AS score]`, // fetch more to allow filtering
            "PARAMS", "2", "blob", blob, "DIALECT", "2");
            // Parse results and filter by user_id if provided
            const results = [];
            for (let i = 1; i < res.length; i += 2) {
                const key = res[i];
                const fields = res[i + 1];
                let id = "";
                let dist = 0;
                let vec_user_id = "";
                for (let j = 0; j < fields.length; j += 2) {
                    if (fields[j] === "id")
                        id = fields[j + 1];
                    if (fields[j] === "score")
                        dist = parseFloat(fields[j + 1]);
                    if (fields[j] === "user_id")
                        vec_user_id = fields[j + 1];
                }
                if (!id)
                    id = key.split(":").pop();
                // Filter by user_id if provided
                if (!user_id || vec_user_id === user_id) {
                    results.push({ id, score: 1 - dist });
                    if (results.length >= topK)
                        break;
                }
            }
            return results;
        }
        catch (e) {
            console.warn(`[Valkey] FT.SEARCH failed for ${sector}, falling back to scan (slow):`, e);
            // Fallback: scan all vectors and filter
            let cursor = "0";
            const allVecs = [];
            do {
                const res = await this.client.scan(cursor, "MATCH", `vec:${sector}:*`, "COUNT", 100);
                cursor = res[0];
                const keys = res[1];
                if (keys.length) {
                    const pipe = this.client.pipeline();
                    keys.forEach(k => pipe.hmget(k, "v", "user_id"));
                    const buffers = await pipe.exec();
                    buffers?.forEach((b, idx) => {
                        if (b && b[1]) {
                            const [buf, vec_user_id] = b[1];
                            const id = keys[idx].split(":").pop();
                            // Filter by user_id during scan
                            if (!user_id || vec_user_id === user_id) {
                                allVecs.push({ id, vector: (0, embed_1.bufferToVector)(buf), user_id: vec_user_id });
                            }
                        }
                    });
                }
            } while (cursor !== "0");
            const sims = allVecs.map(v => ({
                id: v.id,
                score: this.cosineSimilarity(queryVec, v.vector)
            }));
            sims.sort((a, b) => b.score - a.score);
            return sims.slice(0, topK);
        }
    }
    cosineSimilarity(a, b) {
        if (a.length !== b.length)
            return 0;
        let dot = 0, na = 0, nb = 0;
        for (let i = 0; i < a.length; i++) {
            dot += a[i] * b[i];
            na += a[i] * a[i];
            nb += b[i] * b[i];
        }
        return na && nb ? dot / (Math.sqrt(na) * Math.sqrt(nb)) : 0;
    }
    async getVector(id, sector) {
        const key = this.getKey(id, sector);
        const res = await this.client.hmget(key, "v", "dim");
        if (!res[0])
            return null;
        return {
            vector: (0, embed_1.bufferToVector)(res[0]),
            dim: parseInt(res[1])
        };
    }
    async getVectorsById(id) {
        const results = [];
        let cursor = "0";
        do {
            const res = await this.client.scan(cursor, "MATCH", `vec:*:${id}`, "COUNT", 100);
            cursor = res[0];
            const keys = res[1];
            if (keys.length) {
                const pipe = this.client.pipeline();
                keys.forEach(k => pipe.hmget(k, "v", "dim"));
                const res = await pipe.exec();
                res?.forEach((r, idx) => {
                    if (r && r[1]) {
                        const [v, dim] = r[1];
                        const key = keys[idx];
                        const parts = key.split(":");
                        const sector = parts[1];
                        results.push({
                            sector,
                            vector: (0, embed_1.bufferToVector)(v),
                            dim: parseInt(dim)
                        });
                    }
                });
            }
        } while (cursor !== "0");
        return results;
    }
    async getVectorsBySector(sector) {
        const results = [];
        let cursor = "0";
        do {
            const res = await this.client.scan(cursor, "MATCH", `vec:${sector}:*`, "COUNT", 100);
            cursor = res[0];
            const keys = res[1];
            if (keys.length) {
                const pipe = this.client.pipeline();
                keys.forEach(k => pipe.hmget(k, "v", "dim"));
                const res = await pipe.exec();
                res?.forEach((r, idx) => {
                    if (r && r[1]) {
                        const [v, dim] = r[1];
                        const key = keys[idx];
                        const id = key.split(":").pop();
                        results.push({
                            id,
                            vector: (0, embed_1.bufferToVector)(v),
                            dim: parseInt(dim)
                        });
                    }
                });
            }
        } while (cursor !== "0");
        return results;
    }
}
exports.ValkeyVectorStore = ValkeyVectorStore;
