// openmemory-js/dist/server/routes/sources.js
"use strict";
/**
 * sources webhook routes - ingest data from external sources via HTTP
 *
 * POST /sources/:source/ingest
 *   body: { creds: {...}, filters: {...}, user_id?: string }
 *
 * POST /sources/webhook/:source
 *   generic webhook endpoint for source-specific payloads
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.src = src;
const sources = __importStar(require("../../sources"));
function src(app) {
    app.get("/sources", async (req, res) => {
        res.json({
            sources: ["github", "notion", "google_drive", "google_sheets", "google_slides", "onedrive", "web_crawler"],
            usage: {
                ingest: "POST /sources/:source/ingest { creds: {}, filters: {}, user_id? }",
                webhook: "POST /sources/webhook/:source (source-specific payload)"
            }
        });
    });
    app.post("/sources/:source/ingest", async (req, res) => {
        const { source } = req.params;
        const { creds = {}, filters = {}, user_id } = req.body || {};
        const source_map = {
            github: sources.github_source,
            notion: sources.notion_source,
            google_drive: sources.google_drive_source,
            google_sheets: sources.google_sheets_source,
            google_slides: sources.google_slides_source,
            onedrive: sources.onedrive_source,
            web_crawler: sources.web_crawler_source,
        };
        if (!source_map[source]) {
            return res.status(400).json({ error: `unknown source: ${source}`, available: Object.keys(source_map) });
        }
        try {
            const src = new source_map[source](user_id);
            await src.connect(creds);
            const ids = await src.ingest_all(filters);
            res.json({ ok: true, ingested: ids.length, memory_ids: ids });
        }
        catch (e) {
            res.status(500).json({ error: e.message });
        }
    });
    app.post("/sources/webhook/github", async (req, res) => {
        const event_type = req.headers["x-github-event"];
        const payload = req.body;
        if (!payload) {
            return res.status(400).json({ error: "no payload" });
        }
        try {
            const { ingestDocument } = await Promise.resolve().then(() => __importStar(require("../../ops/ingest")));
            let content = "";
            let meta = { source: "github_webhook", event: event_type };
            if (event_type === "push") {
                const commits = payload.commits || [];
                content = commits.map((c) => `${c.message}\n${c.url}`).join("\n\n");
                meta.repo = payload.repository?.full_name;
                meta.ref = payload.ref;
            }
            else if (event_type === "issues") {
                content = `[${payload.action}] ${payload.issue?.title}\n${payload.issue?.body || ""}`;
                meta.repo = payload.repository?.full_name;
                meta.issue_number = payload.issue?.number;
            }
            else if (event_type === "pull_request") {
                content = `[${payload.action}] PR: ${payload.pull_request?.title}\n${payload.pull_request?.body || ""}`;
                meta.repo = payload.repository?.full_name;
                meta.pr_number = payload.pull_request?.number;
            }
            else {
                content = JSON.stringify(payload, null, 2);
            }
            if (content) {
                const result = await ingestDocument("text", content, meta);
                res.json({ ok: true, memory_id: result.root_memory_id, event: event_type });
            }
            else {
                res.json({ ok: true, skipped: true, reason: "no content" });
            }
        }
        catch (e) {
            res.status(500).json({ error: e.message });
        }
    });
    app.post("/sources/webhook/notion", async (req, res) => {
        const payload = req.body;
        try {
            const { ingestDocument } = await Promise.resolve().then(() => __importStar(require("../../ops/ingest")));
            const content = JSON.stringify(payload, null, 2);
            const result = await ingestDocument("text", content, { source: "notion_webhook" });
            res.json({ ok: true, memory_id: result.root_memory_id });
        }
        catch (e) {
            res.status(500).json({ error: e.message });
        }
    });
}
